app/services/git-history.ts

```diff
@@ -0,0 +1,307 @@
+import { Octokit } from '@octokit/rest';
+import { supabase } from '../../src/lib/supabase';
+import { Repository } from '../types/github';
+
+interface FileContributor {
+  file_path: string;
+  contributor_id: string;
+  commit_count: number;
+  last_commit_at: string;
+}
+
+interface CommitFile {
+  filename: string;
+  additions: number;
+  deletions: number;
+  changes: number;
+  status: string;
+}
+
+/**
+ * Index git history for a repository
+ */
+export async function indexGitHistory(
+  repository: Repository,
+  octokit: Octokit,
+  since?: Date
+): Promise<void> {
+  console.log(`Starting git history indexing for ${repository.full_name}`);
+  
+  try {
+    // Get or create repository record
+    const { data: dbRepo } = await supabase
+      .from('repositories')
+      .select('id')
+      .eq('github_id', repository.id)
+      .single();
+    
+    if (!dbRepo) {
+      console.error('Repository not found in database');
+      return;
+    }
+    
+    // Determine the date to start indexing from
+    const sinceDate = since || new Date(Date.now() - 180 * 24 * 60 * 60 * 1000); // 6 months ago
+    
+    // Track file contributors
+    const fileContributors = new Map<string, Map<string, FileContributor>>();
```

Comment:
refactor (suggestion): As mentioned for the embeddings above, the `fileContributors` Map accumulates all file contributors in memory throughout the entire process. For large repos with many files and contributors, this could cause memory issues. Consider processing and flushing data in smaller chunks.